### PR TITLE
Remove regular expressions

### PR DESCRIPTION
**Pull request by jeuxdeau**
As the former version cannot turn off regular expression config in Jerryscript, I removed the codes relating to regular expressions. It can reduce memory usage of apps which do not use regular expressions.

IoT.js-DCO-1.0-Signed-off-by: Woobin Park keyxiz@gmail.com

### CODE REVIEW

File: src/js/ble_uuid_util.js [hunk 1]
```diff
@@ -36,8 +36,36 @@
...
+    while (end < uuid.length) {
+      if (uuid.charAt(end) !== '-') end++;
+      else {
+        str += uuid.slice(start, end);
+        start = end = end + 1;
+      }
+    }
+
+    str += uuid.slice(start, end);
   }
```
Comment (glistening) [inline src/js/ble_uuid_util.js]: I think line 39 ~ line 52 can be replaced with `uuid.split('-').join('')`;

**Comment by glistening** [comment]
@jeuxdeau Could you give me the binary size after disabling jerrsycript regexp with this patch (including base (default build) binary size?

**Comment by jeuxdeau** [comment]
@glistening Thank you, I updated it. And here is it below:

* original: 617.4 kB (617,392 B)
* current: 617.2 kB (617,216 B)
* current(regex disabled): 590.2 kB (590,152 B)

build option: --no-check-test --buildtype release

File: src/js/ble_uuid_util.js [hunk 2]
```diff
@@ -36,8 +36,23 @@
...
 module.exports.removeDashes = function(uuid) {
   if (uuid) {
-    uuid = uuid.replace(/-/g, '');
+    uuid = uuid.split('-').join('');
   }
 
   return uuid;
 };
+
+module.exports.tokenize = function(str) {
```
Comment (glistening) [inline src/js/ble_uuid_util.js]: @jeuxdeau What about `splitAtEvery2ndChars`? I cannot know the expected working by the name of `tokenize`.

File: src/js/ble_uuid_util.js [hunk 3]
```diff
@@ -36,8 +36,23 @@
...
+  var i = 0;
+
+  while (i < len - (len % 2)) {
+    arr.push(str.charAt(i) + str.charAt(i + 1));
+    i = i + 2;
+  }
+
+  if (i < len) arr.push(str.charAt(i));
+  return arr;
+}
```
Comment (zherczeg) [inline src/js/ble_uuid_util.js]: It seems the code always uses .reverse().join('') after this call, perhaps we could do that here.

File: src/js/ble_hci_socket_hci.js [hunk 4]
```diff
@@ -562,7 +563,7 @@ Hci.prototype.processCmdCompleteEvent = function(cmd, status, result) {
     this.emit('readLocalVersion', hciVer, hciRev, lmpVer, manufacturer, lmpSubVer);
   } else if (cmd === READ_BD_ADDR_CMD) {
     this.addressType = 'public';
-    this.address = result.toString('hex').match(/.{1,2}/g).reverse().join(':');
+    this.address = UuidUtil.tokenize(result.toString('hex'), ':');
```
Comment (glistening) [inline src/js/ble_hci_socket_hci.js]: Where can I find the definition of `UuidUtil.tokenize`?

File: src/js/ble_uuid_util.js (hunk 1 above)
Comment (glistening) [inline src/js/ble_uuid_util.js]: @jeuxdeau Could you explain why did you restore the code?

File: src/js/ble_uuid_util.js [hunk 5]
```diff
@@ -36,8 +36,36 @@
...
+    }
+
+    str += uuid.slice(start, end);
   }
 
-  return uuid;
+  return str;
 };
+
+module.exports.refineUuid = function(str, joint) {
```
Comment (glistening) [inline src/js/ble_uuid_util.js]: @jeuxdeau What do you mean by refineUuid? is it general term used in uuid? I can't guess what the function `refineUuid` does by the name.

File: src/js/ble_hci_socket_hci.js (hunk 4 above)
Comment (jeuxdeau) [inline src/js/ble_hci_socket_hci.js]: @glistening I did it by mistake. I'll put it back. Thank you.

File: src/js/ble_uuid_util.js (hunk 1 above)
Comment (jeuxdeau) [inline src/js/ble_uuid_util.js]: @glistening I did it by mistake. I'll put it back. Thank you.

File: src/js/ble_uuid_util.js (hunk 5 above)
Comment (jeuxdeau) [inline src/js/ble_uuid_util.js]: Originally, Tokenize was a function that slices a string into two chars as you suggested. At the advice of @zherczeg, I combined Tokenize with reverse().join(sth). In so doing it got to need more comprehensive name not just representing matching regular expressions but also including its other operations. In this regard, Could you suggest an alternative again? I admit that refine is too ambiguous and not general for uuid as you think.

File: src/js/ble_hci_socket_hci.js [hunk 6]
```diff
@@ -38,6 +38,7 @@ var debug = console.log; // require('debug')('ble_hci');
 
 var events = require('events');
 var util = require('util');
+var UuidUtil = require('ble_uuid_util');
```
Comment (zherczeg) [inline src/js/ble_hci_socket_hci.js]: The otherst start with lowercase letter.

File: src/js/ble_uuid_util.js [hunk 7]
```diff
@@ -36,8 +36,23 @@
...
+  str = String(str);
+  var len = str.length;
+  var arr = [];
+  var i = 0;
+
+  while (i < len - (len % 2)) {
+    arr.push(str.charAt(i) + str.charAt(i + 1));
```
Comment (zherczeg) [inline src/js/ble_uuid_util.js]: what about `str.slice(i, i+1)` ?

File: src/js/ble_uuid_util.js [hunk 8]
```diff
@@ -36,8 +36,23 @@
...
   return uuid;
 };
+
+module.exports.refineUuid = function(str, joint) {
+  str = String(str);
+  var len = str.length;
+  var arr = [];
+  var i = 0;
+
+  while (i < len - (len % 2)) {
```
Comment (zherczeg) [inline src/js/ble_uuid_util.js]: `(len & ~0x1)` ?

File: src/js/ble_uuid_util.js [hunk 9]
```diff
@@ -36,8 +36,23 @@
...
+    arr.push(str.charAt(i) + str.charAt(i + 1));
+    i = i + 2;
+  }
+
+  if (i < len) arr.push(str.charAt(i));
```
Comment (zherczeg) [inline src/js/ble_uuid_util.js]: `len & 0x1` ?

File: src/js/ble_uuid_util.js [hunk 10]
```diff
@@ -36,8 +36,23 @@
...
+  return arr.reverse().join(joint);
```
Comment (zherczeg) [inline src/js/ble_uuid_util.js]: What about creating the string on-the-fly without using an array? Should be faster and use less memory. You can also iterate back to front.

File: src/js/ble_uuid_util.js (hunk 5 above)
Comment (glistening) [inline src/js/ble_uuid_util.js]: @jeuxdeau `splitIntoTwoLetterSegmentsAndReverseAndJoin` or `reverseByteOrder`?

File: src/js/ble_uuid_util.js [hunk 11]
```diff
@@ -36,8 +36,23 @@
...
 module.exports.removeDashes = function(uuid) {
   if (uuid) {
-    uuid = uuid.replace(/-/g, '');
+    uuid = uuid.split('-').join('');
   }
 
   return uuid;
 };
+
+module.exports.refineUuid = function(str, joint) {
```
Comment (glistening) [inline src/js/ble_uuid_util.js]: @jeuxdeau What about `splitIntoTwoLetterSegmentsAndReverseAndJoin` or `reverseByteOrder`?

**Comment by yichoi** [comment]
LGTM

**Pull request merged**